website/docs/wiki/caching.md

```diff
@@ -0,0 +1,84 @@
+
+That's why ProtoCat has a mechanism to remove both of the problems: generic approach for all RPCs without mapping functions and avoid double encoding.
+
+## Cache middleware
+
+If you don't care too much about the details, you can just use
+
+```typescript
+import { createCache, CacheImplementation } from 'protocat'
+
+const cache: CacheImplementation = {
+  set: (key, value, call) => {
+    await someStorage.set(key, value) // MUST SET BUFFER
+  },
+  get: (key, call) => someStorage.get(key), // MUST RETURN BUFFER
+  hash: call =>
+    `${call.path}::${call.user?.id ?? ''}::${call.request.toArray().join('-')}`,
+}
+app.use(
+  createCache(cache, (call, hit, hash) => {
+    call.initialMetadata.set('cache', hit ? 'hit' : 'miss')
+    call.initialMetadata.set('cache-key', hash)
+  })
+)
+```
+
+### Cache storage
+
+It does not matter what implementation you have, any sync/async key-value storage that can handle buffers is fine.
+
+:::danger
+It is crucial that you return the binary data in buffer as you received them. Be mindful of any encoding changes or data truncate that might happen in your cache. If you fail to return a buffer with the same contents as provided by the set function, you will likely send invalid data on the wire. This will likely result in your client receiving "valid" responses but will crash on the data decoding.
+:::
+
+### Different caching settings
+
+In case you don't want to cache some RPC, your `hash` must return falsy value. In that case the middleware will not try to retrieve value from cache nor will save the result. Implement any logic your heart desires in `hash` function. You can read the RPC info, client metadata, data from previous middlewares etc.
+
+### TTL and cache purging
+
+It is all up to you. If you want different TTL for different calls, set the TTL (or delete the record in your cache) yourself. If your `cache.get` returns falsy value, we consider it to be a cache miss for whatever reason (cold start, ttl, etc.). Same goes for cache purging. ProtoCat never needs to clear your cache, so there is no API in the interface for that and you manage it on your own.
+
+## Under the hood
+
+ProtoCat uses the "lowest" available public API of `grpc-js` to hook handlers, that is `server.register`. This allows it to override serialization methods. This way the standard protobuf message to wire function can be overriden to do _noop_ if ProtoCat tells it buffer caching has been used and it skips the default transformation of message to binary.
+
+This is also the reason how we solved the generic method. Now that the wire format can be used directly, it is used as a serialization. That is why it is so crucial to respect the buffer contract and not to tamper with the data in any way. Under normal cricumstances, sending an unserielizable input down gRPC would cause an error on the server, during the encoding. With caching middleware this runtime checks are skipped.
```

Comment:
✏️ I'm having a hard time understanding this paragraph. Is this what you meant?
```suggestion
This is also the reasoning behind how we implemented the generic method. Now that the wire format can be used directly, it is also used as a serialization format that goes to cache storage. That is why it is so crucial to respect the buffer contract and not to tamper with the data in any way. Under normal cricumstances, sending an unserielizable input down gRPC would cause an error on the server, during the encoding. With caching middleware these runtime checks are skipped.
```

Reply:
Cool, thx.